src/anonymous-credentials/presentation-builder.ts

```diff
@@ -492,6 +560,96 @@ export class PresentationBuilder extends Versioned {
       flattenedSchemas.push(flattenedSchema);
     }
 
+    // Create statements and witnesses for each boundedPseudonyms
+    const presentedBoundedPseudonyms: Map<string, IPresentedBoundedPseudonym> = new Map();
+    for (let i = 0; i < this.boundedPseudonyms.length; i++) {
+      // the second parameter is the corresponding attribute index
+      // to do witnessEq.addWitnessRef(sIdx, index);
+      // later when sIdx will be known
+      const witnessEqs: [WitnessEqualityMetaStatement, number][] = [];
+      const attributeBoundPseudonym = this.boundedPseudonyms[i];
+      const attributes: Uint8Array[] = [];
+      for (const [credIdx, attributeNames] of attributeBoundPseudonym.attributeNames.entries()) {
+        // lots of the following is extracted from getRevealedAndUnrevealed and should probably be packaged in a function
+        const cred = this.credentials[credIdx][0];
+        const schema = cred.schema as CredentialSchema;
+        const messages = cred.serializeForSigning();
+        const encoder = schema.encoder;
+        const [names, encodedValues] = encoder.encodeMessageObject(messages);
+        let found = 0;
+        for (let j = 0; j < names.length; j++) {
+          if (attributeNames.has(names[j])) {
+            attributes.push(encodedValues[j]);
+            const witnessEq = new WitnessEqualityMetaStatement();
+            witnessEq.addWitnessRef(credIdx, j);
+            witnessEqs.push([witnessEq, attributes.length - 1]);
+            found++;
+          }
+        }
+        if (attributeNames.size !== found) {
+          throw new Error(
+            `Some of the attribute name for pseudonym were not found in the given messages object, ${
+              attributeNames.size - found
+            } extra names found`
+          );
+        }
+      }
+      const pseudonym = AttributeBoundPseudonym.new(
+        attributeBoundPseudonym.basesForAttributes,
+        attributes,
+        attributeBoundPseudonym.baseForSecretKey,
+        attributeBoundPseudonym.secretKey
+      );
+      const baseForSecretKey = attributeBoundPseudonym.baseForSecretKey;
+      const decodedBaseForSecretKey =
+        baseForSecretKey !== undefined ? PseudonymBases.decodeBaseForSecretKey(baseForSecretKey) : undefined;
+      const decodedBasesForAttributes = PseudonymBases.decodeBasesForAttributes(
+        attributeBoundPseudonym.basesForAttributes
+      );
+      const decodedPseudonym = PseudonymBases.decoder.decode(pseudonym.value);
+      presentedBoundedPseudonyms.set(decodedPseudonym, {
+        commitKey: {
```

Comment:
I don't know what a reference to the params mean. Can you elaborate? What is it? How do you link the reference with the bases?
Do you expect verifier to keep track of the bases they sent to the prover to do the following?:
https://github.com/nicobao/crypto-wasm-ts/blob/6f3fb78322f2fba4bd6d6f1cfbc957cb93f0b00e/src/anonymous-credentials/presentation.ts#L296-L300

Reply:
By reference, I meant a unique (in the pres. context) id of that param. Eg, The snark public key is called `pkId` in the pres. spec [here](https://github.com/nicobao/crypto-wasm-ts/blob/master/tests/anonymous-credentials/presentation.spec.ts#L961) and the verifier uses it like [this](https://github.com/nicobao/crypto-wasm-ts/blob/master/tests/anonymous-credentials/presentation.spec.ts#L968). This was to avoid creating multiple copies of a param object if the same param is used in multiple statements. But now i think what you have is fine because the bases wont be reused in multiple statements and these are not as big as snark keys so its fine.